### ISSUE TITLE
BUG: int to np.float32 conversion is not correctly rounded

### ISSUE THREAD

**Issue by mdickinson** [numpy/numpy#20687]
### Describe the issue:

Conversion from a Python `int` to an `np.float32` value does not appear to be correctly rounded, and does not always match the equivalent conversion from `np.int64` to `np.float32`.

I'm a bit hesitant to call this a bug rather than a feature request, since I'm not aware of anywhere where it's specified that `int` to `np.float32` conversions _should_ be correctly rounded. But whether bug or not, correct rounding seems like a desirable property here, and consistency between `int` and `np.int64` also seems desirable.

In IEEE 754 binary32 floating-point, `2**54 - 2**30` and `2**54` are both exactly representable, and are neighboring floating-point values. The halfway point between them is `2**54 - 2**29`. Under correct rounding (ties-to-even, as usual), integers `n` in the interval `[2**54 - 2**29, 2**54)` should round up to `2**54` when converted to `np.float32`, while integers in `(2**54-2**30, 2**54-2**29)` should round down to `2**54 - 2**30`. In particular, `2**54 - 2**29 - 1` should round down, but as the code example below shows, it rounds up.

In contrast, the exact same input value represented as an `int64` instead of an `int` rounds down (as expected).

I haven't looked at the implementation, but my strong suspicion is that there's a double rounding going on here: that the incoming `int` is converted first to `float64`, and then to `float32`.

### Reproduce the code example:

```python
>>> import numpy as np
>>> np.float32(2**54 - 2**29 - 1) == 2**54 - 2**30  # expect True, get False
False
>>> np.float32(np.int64(2**54 - 2**29 - 1)) == 2**54 - 2**30  # ok: expect True, get True
True
```

### NumPy/Python version information:

```
>>> import sys, numpy
>>> print(numpy.__version__)
1.21.4
>>> print(sys.version)
3.9.9 (main, Nov 16 2021, 07:23:17) 
[Clang 10.0.1 (clang-1001.0.46.4)]
```

**Comment by seberg** [comment]
I would consider this a bug, although maybe a small one.  The thing is that normal casting of course works correctly (as it uses the C cast).  But coercing from a scalar runs into the issue.

That means we have a different result depending on whether we assign from an array or a scalar and that is an inconsistency.

**Comment by mdickinson** [comment]
Yes, I'd expect no-one to notice or care about the lack of correct rounding, and that's fine. The inconsistency between `int` and `int64` seems more troublesome.

I took a brief look at the correct rounding issue: Python's C-API doesn't make it at all easy to do a correct rounding from a `Py_LongObject` to a (presumed IEEE 754 binary32) C `float` *efficiently*. (Though it's easy enough if you don't care about efficiency.)

I wonder whether a simple "try converting the PyLong to int64 or uint64 first" approach would be good enough: it would resolve the consistency issue, but not the correct rounding one.
